File: sdk-dotnet/src/Inferable.cs
using System.Text.Json.Serialization;
using Inferable.API;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Inferable
{
  public class Links
  {
    public static string DOCS_AUTH = "https://docs.inferable.ai/pages/auth";
  }

  /// <summary>
  /// Input object for onStatusChange functions
  /// Generic type T is the type of the result the run's (resultSchema)
  /// https://docs.inferable.ai/pages/runs#onstatuschange
  /// </summary>
  public struct OnStatusChangeInput<T>
  {
    [JsonPropertyName("runId")]
    public string RunId { get; set; }

    [JsonPropertyName("status")]
    public string Status { get; set; }

    [JsonPropertyName("result")]
    public T? Result { get; set; }

    [JsonPropertyName("tags")]
    public Dictionary<string, string> Tags { get; set; }
  }

  /// <summary>
  /// Input object for handleCustomAuth functions
  /// https://docs.inferable.ai/pages/custom-auth
  /// </summary>
  public struct HandleCustomAuthInput
  {
    [JsonPropertyName("token")]
    public string Token { get; set; }
  }

  public class InferableOptions
  {
    public string? BaseUrl { get; set; }
    public string? ApiSecret { get; set; }
    public string? MachineId { get; set; }
  }

  public struct PollRunOptions
  {
    public required TimeSpan MaxWaitTime { get; set; }
    public required TimeSpan Interval { get; set; }
  }


  /// <summary>
  /// The Inferable client. This is the main entry point for using Inferable.
  /// Basic usage:
  /// <code>
  /// // create a new Inferable instance
  /// var client = new InferableClient(new InferableOptions {
  ///     ApiSecret = "API_SECRET"
  /// });
  /// </code>
  /// </summary>
  public class InferableClient
  {
    public static string DefaultBaseUrl = "https://api.inferable.ai/";

    private readonly ApiClient _client;
    private readonly ILogger<InferableClient> _logger;
    private string? _clusterId;

    private List<IToolRegistration> _toolRegistry = new List<IToolRegistration>();

    private List<PollingAgent> _pollingAgents = new List<PollingAgent>();

    /// <summary>
    /// Initializes a new instance of the InferableClient class.
    /// Basic usage:
    /// <code>
    /// // Create a new Inferable instance with an API secret
    /// var client = new InferableClient(new InferableOptions {
    ///     ApiSecret = "API_SECRET"
    /// });
    ///
    /// // OR
    ///
    /// Environment.SetEnvironmentVariable("INFERABLE_API_SECRET", "API_SECRET");
    /// var client = new InferableClient();
    /// </code>
    /// </summary>
    public InferableClient(InferableOptions? options = null, ILogger<InferableClient>? logger = null)
    {
      string? apiSecret = options?.ApiSecret ?? Environment.GetEnvironmentVariable("INFERABLE_API_SECRET");
      string baseUrl = options?.BaseUrl ?? Environment.GetEnvironmentVariable("INFERABLE_API_ENDPOINT") ?? DefaultBaseUrl;
      string machineId = options?.MachineId ?? Machine.GenerateMachineId();

      if (apiSecret == null)
      {
        throw new ArgumentNullException($"No API Secret provided. Please see ${Links.DOCS_AUTH}");
      }

      if (!apiSecret.StartsWith("sk_"))
      {
        throw new ArgumentException($"Invalid API Secret. Please see: {Links.DOCS_AUTH}");
      }


      this._client = new ApiClient(new ApiClientOptions{
          ApiSecret = apiSecret,
          BaseUrl = baseUrl,
          MachineId = machineId
      });

      this._logger = logger ?? NullLogger<InferableClient>.Instance;
    }

    public void RegisterTool<T>(ToolRegistration<T> tool) where T : struct {
      var existing = this._toolRegistry.FirstOrDefault(f => f.Name == tool.Name);
      if (existing != null) {
        throw new Exception($"Function with name '{tool.Name}' already registered");
      }

      this._toolRegistry.Add(tool);
    }

    public async Task ListenAsync() {
      if (this._pollingAgents.Count > 0) {
        throw new Exception("Already polling");
      }

      if (this._toolRegistry.Count == 0) {
        throw new Exception($"No functions registered");
      }


      var clusterId = await this.GetClusterId();
      var agent = new PollingAgent(clusterId, this._client, this._logger, this._toolRegistry);

      await agent.Start();
      this._pollingAgents.Add(agent);
    }


    public async Task UnListenAsync() {
      if (this._pollingAgents.Count == 0) {
        throw new Exception("Not polling");
      }

      foreach (var agent in this._pollingAgents) {
        await agent.Stop();
      }

      this._pollingAgents.Clear();
    }

    internal async Task<string> GetClusterId() {
      if (this._clusterId == null) {
        // Call register machine without any services to test API key and get clusterId
        var registerResult = await _client.CreateMachine(new CreateMachineInput {});
        this._clusterId = registerResult.ClusterId;
      }

      return this._clusterId;
    }
  }
}

File: sdk-node/src/Inferable.ts
import debug from "debug";
import path from "path";
import { z } from "zod";
import { createApiClient } from "./create-client";
import { InferableAPIError, InferableError } from "./errors";
import * as links from "./links";
import { machineId } from "./machine-id";
import { PollingAgent, registerMachine } from "./polling";
import {
  ContextInput,
  ToolConfig,
  ToolInput,
  ToolRegistrationInput,
  JsonSchemaInput,
} from "./types";
import { helpers, Workflow } from "./workflows/workflow";

// Custom json formatter
debug.formatters.J = (json) => {
  return JSON.stringify(json, null, 2);
};

export const log = debug("inferable:client");

/**
 * The Inferable client. This is the main entry point for using Inferable.
 *
 * ```ts
 * // create a new Inferable instance
 * const client = new Inferable({
 *  apiSecret: "API_SECRET",
 * });
 *
 *
 * // Register a tool
 * client.tools.register("hello", z.object({name: z.string()}), async ({name}: {name: string}) => {
 *  return `Hello ${name}`;
 * })
 *
 * await client.tools.listen();
 *
 * // stop the service on shutdown
 * process.on("beforeExit", async () => {
 *   await myService.stop();
 * });
 *
 * ```
 */
export class Inferable {
  static getVersion(): string {
    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
    return require(path.join(__dirname, "..", "package.json")).version;
  }

  private clusterId?: string;

  private apiSecret: string;
  private endpoint: string;
  private machineId: string;

  private client: ReturnType<typeof createApiClient>;

  private pollingAgents: PollingAgent[] = [];

  private toolsRegistry: { [key: string]: ToolRegistrationInput<any> } = {};

  /**
   * Initializes a new Inferable instance.
   * @param apiSecret The API Secret for your Inferable cluster. If not provided, it will be read from the `INFERABLE_API_SECRET` environment variable.
   * @param options Additional options for the Inferable client.
   * @param options.endpoint The endpoint for the Inferable cluster. Defaults to https://api.inferable.ai.
   *
   * @example
   * ```ts
   * // Basic usage
   * const client = new Inferable({
   *  apiSecret: "API_SECRET",
   * });
   *
   * // OR
   *
   * process.env.INFERABLE_API_SECRET = "API_SECRET";
   * const client = new Inferable();
   * ```
   */
  constructor(options?: {
    apiSecret?: string;
    endpoint?: string;
    machineId?: string;
  }) {
    if (options?.apiSecret && process.env.INFERABLE_API_SECRET) {
      log(
        "API Secret was provided as an option and environment variable. Constructor argument will be used.",
      );
    }

    const apiSecret = options?.apiSecret || process.env.INFERABLE_API_SECRET;

    if (!apiSecret) {
      throw new InferableError(
        `No API Secret provided. Please see ${links.DOCS_AUTH}`,
      );
    }

    if (!apiSecret.startsWith("sk_")) {
      throw new InferableError(
        `Invalid API Secret. Please see ${links.DOCS_AUTH}`,
      );
    }

    this.apiSecret = apiSecret;

    this.endpoint =
      options?.endpoint ||
      process.env.INFERABLE_API_ENDPOINT ||
      "https://api.inferable.ai";

    this.machineId = options?.machineId || machineId();


    this.client = createApiClient({
      baseUrl: this.endpoint,
      machineId: this.machineId,
      apiSecret: this.apiSecret,
    });
  }

  public tools = {
    /**
     * Registers a tool with Inferable.
     * @param input The tool definition.
     * @example
     * ```ts
     * const client = new Inferable({apiSecret: "API_SECRET"});
     *
     * client.tools.register("hello", z.object({name: z.string()}), async ({name}: {name: string}) => {
     *   return `Hello ${name}`;
     * });
     *
     * // start the service
     * await client.tools.listen();
     *
     * // stop the service on shutdown
     * process.on("beforeExit", async () => {
     *   await client.tools.stop();
     * });
     * ```
     */
    register: <T extends z.ZodTypeAny | JsonSchemaInput>(
      input: ToolRegistrationInput<T>,
    ) => {
      this.registerTool({
        name: input.name,
        description: input.description,
        func: input.func,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        inputSchema: input.schema?.input ?? (z.object({}).passthrough() as any),
        config: input.config,
      });
    },
    listen: async () => {
      if (this.pollingAgents.length > 0) {
        throw new InferableError("Tools already listening");
      }

      // TODO: Create one polling agent per 10 tools
      const agent = new PollingAgent({
        endpoint: this.endpoint,
        machineId: this.machineId,
        apiSecret: this.apiSecret,
        clusterId: await this.getClusterId(),
        tools: Object.values(this.toolsRegistry),
      });

      this.pollingAgents.push(agent);
      await agent.start();
    },
    unlisten: async () => {
      Promise.all(this.pollingAgents.map((agent) => agent.stop()));
    },
  };

  private registerTool<T extends z.ZodTypeAny | JsonSchemaInput>({
    name,
    func,
    inputSchema,
    config,
    description,
  }: {
    name: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    func: (input: ToolInput<T>, context: ContextInput) => any;
    inputSchema: T;
    config?: ToolConfig;
    description?: string;
  }) {
    if (this.toolsRegistry[name]) {
      throw new InferableError(`Tool name '${name}' is already registered.`);
    }

    const registration: ToolRegistrationInput<T> = {
      name,
      func,
      schema: {
        input: inputSchema,
      },
      config,
      description,
    };

    const existing = this.pollingAgents.length > 0;

    if (existing) {
      throw new InferableError(
        `Tools must be registered before starting the listener. Please see ${links.DOCS_FUNCTIONS}`,
      );
    }

    if (typeof registration.func !== "function") {
      throw new InferableError(
        `func must be a function. Please see ${links.DOCS_FUNCTIONS}`,
      );
    }

    log(`Registering tool`, {
      name: registration.name,
    });

    this.toolsRegistry[registration.name] = registration;
  }

  private async getClusterId() {
    if (!this.clusterId) {
      // Call register machine without any services to test API key and get clusterId
      const registerResult = await registerMachine(this.client);
      this.clusterId = registerResult.clusterId;
    }

    return this.clusterId;
  }

  workflows = {
    helpers,
    create: <TInput extends z.ZodTypeAny>({
      name,
      inputSchema,
    }: {
      name: string;
      inputSchema: TInput;
    }) => {
      return new Workflow({
        name,
        inferable: this,
        getClusterId: async () => await this.getClusterId(),
        client: this.client,
        inputSchema,
        endpoint: this.endpoint,
        machineId: this.machineId,
        apiSecret: this.apiSecret,
      });
    },
    trigger: async <TWorkflowInput extends { executionId: string }>(
      name: string,
      input: TWorkflowInput,
    ): Promise<void> => {
      const clusterId = await this.getClusterId();

      const result = await this.client.createWorkflowExecution({
        params: {
          clusterId,
          workflowName: name,
        },
        body: input,
      });

      if (result.status !== 201) {
        throw new InferableAPIError(
          "Failed to create workflow execution",
          result,
        );
      }
    },
  };
}
